src/target/stm32l0.c

```diff
@@ -326,10 +326,9 @@ static bool stm32lx_nvm_busy_wait(target *t, uint32_t nvm)
 	uint32_t sr;
 	do {
 		sr = target_mem_read32(t, STM32Lx_NVM_SR(nvm));
-		if ((sr & STM32Lx_NVM_SR_ERR_M) || !(sr & STM32Lx_NVM_SR_EOP) || target_check_error(t))
-			return false;
 	} while (sr & STM32Lx_NVM_SR_BSY);
-
+	if ((sr & STM32Lx_NVM_SR_ERR_M) || !(sr & STM32Lx_NVM_SR_EOP) || target_check_error(t))
+		return false;
```

Comment:
There are a couple of problems with this change that we see - the first is that `target_check_error(t)` must be called every loop otherwise the loop might never terminate, and the second is to do with the if-return logic (clang-tidy warns not do this!)

We would suggest this should either be:
```c
uint32_t sr;
do {
    sr = target_mem_read32(t, STM32Lx_NVM_SR(nvm));
    if ((sr & STM32Lx_NVM_SR_ERR_M) || target_check_error(t))
        return false;
} while ((sr & STM32Lx_NVM_SR_BSY) && !(sr & STM32Lx_NVM_SR_EOP));
```
with the original post-while return, or.. to keep with the theme here:
```c
while (target_mem_read32(t, STM32Lx_NVM_SR(nvm)) & STM32Lx_NVM_SR_BSY) {
    if (target_check_error(t))
        return false;
}
const uint32_t sr = target_mem_read32(t, STM32Lx_NVM_SR(nvm));
return !target_check_error(t) && !(sr & STM32Lx_NVM_SR_ERR_M) && (sr & STM32Lx_NVM_SR_EOP);
```

The former method keeps the original flow and intent of this code in tact, the latter might actually be slightly more correct and definitely solves your issue still.

Reply:
I think that other SR bits are only valid after BSY is released. Look at my proposal in the updated PR.

Reply:
Up to 91d1ef8bf6abc5f5ffff1 it was:
        do {
                sr = target_mem_read32(t, STM32Lx_NVM_SR(nvm));
        } while (sr & STM32Lx_NVM_SR_BSY);

        if ((sr & STM32Lx_NVM_SR_ERR_M) || !(sr & STM32Lx_NVM_SR_EOP) ||
            target_check_error(t))
                        return -1;

Reply:
This looks better now, but it's still doing a pessimistic do-while loop which is what we're trying to suggest you avoid in the original review above. Otherwise it looks ok. That leaves one last review point before we can merge:

This is still doing `if (cond) return false; return true;` which clang-tidy will tell you not to do (so don't!) - the condition of the if statement should be turned into the return value directly with a single return - `return !(cond);`, so in this case `return !(sr & STM32Lx_NVM_SR_ERR_M) && (sr & STM32Lx_NVM_SR_EOP);`